### ISSUE TITLE
Support HDFS for Storage Profile

### ISSUE THREAD

**Issue by swrookie** [lakekeeper/lakekeeper#485]
According to the current status in the readme, it seems only cloud storages are supported for the catalog.
For companies running on premise data lake, HDFS is widely used.

It would be nice if HDFS is supported as one of the options for Storage Profile.

**Comment by c-thiel** [comment]
Thanks @swrookie for opening this issue.

We have been thinking about HDFS as well. We have been focusing on S3 for on-prem deployments, as just-in-time security mechanism such as "vended-credentials" are currently not supported by clients with HDFS. We would thus lack features for HDFS.

Is HDFS a hard requirement for you? We see more and more companies switching to S3.

I am leaving this issue open to track interest.

**Comment by uniqueinput** [comment]
@c-thiel 
There might be a chance to use HDFS delegation tokens for that purpose
https://blog.cloudera.com/hadoop-delegation-tokens-explained/

**Comment by swrookie** [comment]
Hi @c-thiel.
I personally want to switch to object storage, but currently HDFS is a hard requirement for me.
Wonder if insecure access as part 1 of feature before security feature(vended credentials, delegation tokens) as part 2 would make sense.

**Comment by c-thiel** [comment]
Good news @swrookie, @uniqueinput, we decided to add HDFS.
Following @swrookie approach, we will start without security / vended creds / delegation tokens.

When the PR is ready it would be great if you could have a look if we missed some authentication mechanism that is super important to you.

**Comment by bentzimortaboola** [comment]
we also need support for HDFS
many large on-prem companies using very large HDFS clusters

**Comment by c-thiel** [comment]
@swrookie , @uniqueinput, @bentzimortaboola   and everyone else passing by, do you need the following?
- Is encryption at REST / KMS required?
- Which authentication is used? Kerberos (GSSAPI SASL)? Token (DIGEST-MD5 SASL)? anything else?

**Comment by c-thiel** [comment]
Copied from Discord:

- We need when using webhdfs
- Kerberos

**Comment by bentzimortaboola** [comment]
@c-thiel for us ATM encryption is not required

**Comment by swrookie** [comment]
@c-thiel currently, secure access is not required for us

**Comment by bentzimortaboola** [comment]
@c-thiel also for us , ATM we care about authorization

**Comment by c-thiel** [comment]
@swrookie , @bentzimortaboola, @uniqueinput  thank for your Feedback!

A preview version of HDFS is ready!
HDFS is supported in the following image: quay.io/lakekeeper/catalog:v0.8.3-hdfs-preview
A Docker compose example (with kerberos) is available here: https://github.com/lakekeeper/lakekeeper/tree/ct/hdfs/examples/hdfs
Additional docs are available here: https://github.com/lakekeeper/lakekeeper/blob/ct/hdfs/docs/docs/storage.md#hdfs-storage

Please let us know if you miss anything or encounter problems :).
Once we have enough Feedback, we will merge support into main!

**Comment by bentzimortaboola** [comment]
thanks @c-thiel 
we are not using kerberos authentication in HDFS, will this preview image will work for us?

**Comment by rotem-ad** [comment]
I tested the preview version without kerberos authentication (modified docker compose file is attached). I successfully created a warehouse and a new namespace.

In addition, I was able to connect Trino to the catalog and create, query and drop tables (Trino catalog properties file is attached as well).

Some potential issues I've encountered after performing some basic tests:
1. After creating the warehouse, this directory tree was created with no actual files stored in the metadata dir: `/iceberg_tests/warehouse/01967749-c78a-7a50-b35b-bc6eb9652344/01967749-c78a-7a50-b35b-bc707297c134/metadata` (key-prefix="/iceberg_tests/warehouse"). No files were created after creating Iceberg tables as well. Not sure if this is an expected behavior.
2. In the UI, when clicking the "Connect Compute" in the warehouse menu and then select "Trino", the following error message is shown: `"The warehouse storage profile (hdfs) is not supported by Trino"`. Not really sure why since Trino basic operations worked for me, but I assume I haven't yet discovered the unsupported Trino features. Are you planning to fully support Trino as well?
3. After dropping my tests tables in Trino, I've noticed that the tables' directory structure hasn't been removed from HDFS. Only the actual data/metadata files were deleted from HDFS. Is this an expected behavior?

[docker-compose.yaml.txt](https://github.com/user-attachments/files/19930589/docker-compose.yaml.txt)

[iceberg_rest.properties.txt](https://github.com/user-attachments/files/19930661/iceberg_rest.properties.txt)

**Comment by twuebi** [comment]
Thanks for testing the preview @rotem-ad 

I'll look into your points. Regarding 2., I think this is most likely a false negative in the UI.

**Comment by twuebi** [comment]
> 1. After creating the warehouse, this directory tree was created with no actual files stored in the metadata dir: /iceberg_tests/warehouse/01967749-c78a-7a50-b35b-bc6eb9652344/01967749-c78a-7a50-b35b-bc707297c134/metadata (key-prefix="/iceberg_tests/warehouse"). No files were created after creating Iceberg tables as well. Not sure if this is an expected behavior.

These are remnants of us testing access to the object store, we're creating and deleting a file before issuing a remove_all for the test-location. The library we're using, `hdfs-native-object-store`, has two delete methods (https://docs.rs/object_store/0.12.0/object_store/trait.ObjectStore.html#method.delete_stream, https://docs.rs/object_store/0.12.0/object_store/trait.ObjectStore.html#tymethod.delete), we're using a list of table location followed by a delete_stream. Since the list apparently only includes files and no directories, we end up keeping the directory tree. 

Trying to issue a delete fails with `org.apache.hadoop.fs.PathIsNotEmptyDirectoryException `/user/hdfs4/01967ca1-1aa8-74c2-aafe-e0593cccfc5c/01967ca1-1aa8-74c2-aafe-e06023aeb4b8 is non empty`. Delete does not have a recursive parameter. I'll try if we can use deletes issued inside out.

> 3. After dropping my tests tables in Trino, I've noticed that the tables' directory structure hasn't been removed from HDFS. Only the actual data/metadata files were deleted from HDFS. Is this an expected behavior?

This is the same case as 1.

**Comment by twuebi** [comment]
Since we are only using `object_store` for hdfs, we can probably drop it and directly depend on `hdfs_native` which would expose the parameters we're looking for.

**Comment by bentzimortaboola** [comment]
Hi

Is it possible to send the iceberg.rest-catalog.warehouse in the iceberg.rest-catalog.uri ?
we are working with Vertica analytic database and ask them to check adding support to LakeKeeper, we shared how to connect and they reply that rest catalog is supported but why need to specify the warehouse?

they shared example from aws:
create external table rest_t1 STORED BY ICEBERG LOCATION 'http://servers1.amazonaws.com:19120/iceberg/v1/main/namespaces/my_ice_db/tables/t1' REST_AUTH '{"bearerToken": "$token"}';

Thanks

**Comment by c-thiel** [comment]
> Hi
> 
> Is it possible to send the iceberg.rest-catalog.warehouse in the iceberg.rest-catalog.uri ? we are working with Vertica analytic database and ask them to check adding support to LakeKeeper, we shared how to connect and they reply that rest catalog is supported but why need to specify the warehouse?
> 
> they shared example from aws: create external table rest_t1 STORED BY ICEBERG LOCATION 'http://servers1.amazonaws.com:19120/iceberg/v1/main/namespaces/my_ice_db/tables/t1' REST_AUTH '{"bearerToken": "$token"}';
> 
> Thanks

Specifying a Warehouse is part of the Iceberg REST Specification. If this is not supported by Vertica, then Vertica is not Iceberg REST compliant: https://github.com/apache/iceberg/blob/61e8acecf512d8dd2a72727803e5836ea1099eed/open-api/rest-catalog-open-api.yaml#L336C3-L336C38

Nonetheless, you can do the same with Lakekeeper - you can use the following URL:
`https://<lakekeeper-base-url>/catalog/v1/<id-of-your-warehouse>/namespaces/my_ice_db/tables/t1`. You can get the Warehouse ID from the UI or from the /config Endpoint (called prefix there).

If you need more information, please create a new issue. I don't see how this is connected to HDFS.

**Comment by c-thiel** [comment]
An updated HDFS image is available! `quay.io/lakekeeper/catalog:v0.8.4-hdfs-preview`
Remove all should now work as expected. @rotem-ad your points 1) and 3) are fixed :)

Regarding your point 2), we will ship connect support in the UI in a next release. It's not straight forward though and very much depends on your security setup. Kerberos is just a big mess unfortunately. We might just point to Markdown documentation.

**Comment by MirerRon** [comment]
Hi, first of all thanks for the HDFS integration! 
I was checking the preview against an HDFS cluster (not the one deployed in the compose) and made sure to replace all the site files. But am encountering an error when attempting to create a warehouse. I send the following request (as stated in the guide):
```
{
    "warehouse-name": "hdfs-test",
     "storage-profile": {
            "type": "hdfs",
            "key-prefix": "/lakekeeper-test/hdfs-test",
            "url": "hdfs://Nameservice1",
      }
}
```

And I get the following error: 
``` 
Failed to write file. Please check the storage credentials. Caused by: data transfer error: No new datanode found in updated block locations
```
But it does seem that a test table is created in the warehouse directory. Any idea what may cause it?

**Comment by rotem-ad** [comment]
Thanks @c-thiel!
I tested the new HDFS image and can confirm "remove all" functionality now works as expected (:
Regarding Kerberos, the HDFS cluster I use for testing isn't configured with Kerberos authentication so I can't provide feedback about Kerberos.

On a different note, I've encountered a new issue with Trino integration:
When running `show create schema <iceberg_catalog>.<schema_name>` in Trino, I get this error message: `io.trino.spi.TrinoException: No PropertyMetadata for property: namespace_id`
It looks similar to this other Trino bug: https://github.com/trinodb/trino/issues/24744
I assume this is purely a Trino bug and probably not related to HDFS implementation, but just wanted to bring this to your attention as well.

**Comment by c-thiel** [comment]
> Hi, first of all thanks for the HDFS integration! I was checking the preview against an HDFS cluster (not the one deployed in the compose) and made sure to replace all the site files. But am encountering an error when attempting to create a warehouse. I send the following request (as stated in the guide):
> 
> ```
> {
>     "warehouse-name": "hdfs-test",
>      "storage-profile": {
>             "type": "hdfs",
>             "key-prefix": "/lakekeeper-test/hdfs-test",
>             "url": "hdfs://Nameservice1",
>       }
> }
> ```
> 
> And I get the following error:
> 
> ```
> Failed to write file. Please check the storage credentials. Caused by: data transfer error: No new datanode found in updated block locations
> ```
> 
> But it does seem that a test table is created in the warehouse directory. Any idea what may cause it?

That is interesting, it seems that validation started successfully, but then a later validation step failed.
Could you try to run Lakekeeper with the following environment variable present:
`RUST_LOG=iceberg_catalog=info` and check the logs on the server? You should see the full HDFS error message there - I am not sure though if we see more than `No new datanode found in updated block locations` - but its worth a shot.

**Comment by c-thiel** [comment]
> Thanks [@c-thiel](https://github.com/c-thiel)! I tested the new HDFS image and can confirm "remove all" functionality now works as expected (: Regarding Kerberos, the HDFS cluster I use for testing isn't configured with Kerberos authentication so I can't provide feedback about Kerberos.
> 
> On a different note, I've encountered a new issue with Trino integration: When running `show create schema <iceberg_catalog>.<schema_name>` in Trino, I get this error message: `io.trino.spi.TrinoException: No PropertyMetadata for property: namespace_id` It looks similar to this other Trino bug: [trinodb/trino#24744](https://github.com/trinodb/trino/issues/24744) I assume this is purely a Trino bug and probably not related to HDFS implementation, but just wanted to bring this to your attention as well.

Yeah, this is a trino bug. We are also tracing it here: https://github.com/lakekeeper/lakekeeper/issues/1041

**Comment by twuebi** [comment]
> Hi, first of all thanks for the HDFS integration! I was checking the preview against an HDFS cluster (not the one deployed in the compose) and made sure to replace all the site files. But am encountering an error when attempting to create a warehouse. I send the following request (as stated in the guide):
> 
> ```
> {
>     "warehouse-name": "hdfs-test",
>      "storage-profile": {
>             "type": "hdfs",
>             "key-prefix": "/lakekeeper-test/hdfs-test",
>             "url": "hdfs://Nameservice1",
>       }
> }
> ```
> 
> And I get the following error:
> 
> ```
> Failed to write file. Please check the storage credentials. Caused by: data transfer error: No new datanode found in updated block locations
> ```
> 
> But it does seem that a test table is created in the warehouse directory. Any idea what may cause it?

Hi @MirerRon, googling the error brought this location within hdfs_native up: https://docs.rs/hdfs-native/latest/src/hdfs_native/hdfs/block_writer.rs.html#691 which seems to happen within the unhappy path of a write, it's getting called from within `recover` https://docs.rs/hdfs-native/latest/src/hdfs_native/hdfs/block_writer.rs.html#397-400. As @c-thiel said, running the server with `RUST_LOG=debug` and sharing the logs with us can help us pinpoint the issue.

**Comment by MirerRon** [comment]
Thanks for the suggestion @twuebi @c-thiel. Using RUST_LOG=debug still only raised this error. I saw the function, and had a thought that it might have to do with something with the cluster I was using and the amount of nodes. Moved to another HDFS cluster and I get a different error: "Datanode connection closed while for ack", which happens when the server attempts to delet ethe test table metadata directory. Seems odd considering it's able to create the warehouse directory and the metadata file itself (Which contains "test" as I guess is expected). 

I looked a bit in the hdfs_native client, and I think it can be the forced usage of SASL authentication (while I do use kerberos the SASL configuration is not turned on in my cluster). I saw that it forces a whole bunch of configurations as their default value apart from a small required batch, which is not ideal :( considering there are configurations we would like to be applied for clients. I would try to open an issue in the hdfs_native repo.

**Comment by Kimahriman** [comment]
Author of both `hdfs-native` and `hdfs-native-object-store` here, let me know any other issues you encounter or any client configs you rely on, happy to help and make improvements where possible

**Comment by Kimahriman** [comment]
One minor thing in the docs:

>Lakekeeper supports using Hadoop Distributed File System (HDFS) as a storage location for Iceberg tables. This integration is built on the [hdfs-native](https://crates.io/crates/hdfs-native) crate, which provides native Rust bindings to libhdfs3 for high-performance HDFS access without requiring a JVM.

It's not bindings on top of `libhdfs3`, it's just a pure Rust implementation from scratch.

**Comment by MirerRon** [comment]
@@twuebi @c-thiel , I am bit confused beacuse I have a log "Failed to delete table metadata from '/lakekeeper/hdfs-test-2/\<ns-uuid\>/\<tbl-uuid\>/metadata/test'". I think this is probably a mistake beacuse this log is raised from the `write_metadata_file()` function that is called in the `validate_read_write()` function. Probably just a mistake in the [log](https://github.com/lakekeeper/lakekeeper/blob/ct/hdfs/crates/iceberg-catalog/src/catalog/io/mod.rs#L68) that should be looked at? But I am yet to understand why another file is written? beacuse the `.test.tmp.1` file already exists (couldn't find the function that creates it), so why does another "test" file need to be created?

Hoping you can help me a bit here so I'll be able to cross-reference with the `hdfs-native` client and might pinpoint my issue. Thanks!

**Comment by c-thiel** [comment]
@MirerRon  the `.test.tmp.1` file comes from here: https://github.com/lakekeeper/lakekeeper/blob/6ad72388c4b5e295df1db3159c1fe1ab100851c4/crates/iceberg-catalog/src/catalog/io/hdfs.rs#L39

Atomtic writes to HDFS work by writing a tmp file first, then renaming it.
If the test file is there, then maybe only renaming fails?
Could you try to follow these steps with another client?

**Comment by c-thiel** [comment]
@MirerRon you are right, the log is wrong. If fixed it and also added new logs that might help to debug the issue.
A new image tag is now available: `quay.io/lakekeeper/catalog:v0.8.5-hdfs-preview`
It might make sense to use `RUST_LOG=iceberg_catalog=trace`

@Kimahriman thanks for the note on the docs! I fixed the comment now.
Thanks for your great crate - it really helps to have this rust native!

@MirerRon  could you post some of your logs with the latest release if the problem persists?

**Comment by boroknagyz** [comment]
Thank you for working on this! For Apache Impala we are currently working on Iceberg REST Catalog support. In our development environment we typically have an HDFS cluster to test with, so your work makes it easy for us to test against Lakekeeper. So far so good!

**Comment by rdsarvar** [comment]
Out of curiosity @c-thiel, where has this landed? Is it mostly pending @MirerRon trying with the newest changes? Or are you looking for more feedback?

**Comment by c-thiel** [comment]
@rdsarvar I haven't picked up the topic since and request for HDFS is currently low.
We might pick up the topic in the future but are currently not investing more time on it.
Community contributions are welcome though!
